Improve test coverage for errors

betasve opened the pull request:
Signed-off-by: Svetoslav Blyahov <svetlio.blyahoff@gmail.com>

Improve configtxgen test coverage (for error messages) as @wlahti advised in my previous PR.

#### Type of change
- Test update

#### Description

Start expecting exact error messages.

noamher commented:
Signed-off-by: noam@digitalasset.com

srderson commented:
LGTM. Thanks!

betasve committed:
Improve test coverege for errors

Signed-off-by: Svetoslav Blyahov <svetlio.blyahoff@gmail.com>

wlahti commented on cmd/configtxgen/main_test.go:
```diff
@@ -77,7 +78,20 @@ func TestMissingConsortiumValue(t *testing.T) {
+
+	err := doOutputChannelCreateTx(config, nil, "foo", configTxDest).Error()
+	assert.Error(t, doOutputChannelCreateTx(config, nil, "foo", configTxDest), "Not able to write to file")
+	assert.Regexp(t, regexp.MustCompile("Error writing channel create tx: open /tmp/configtxgen.+/configtx: permission denied"), err)
```
This works in CI but won't work locally depending on the developer's OS (i.e. it fails for me on mac).

Also a nit: you can leave off regexp.MustCompile

wlahti commented on cmd/configtxgen/main_test.go:
```diff
@@ -77,7 +78,20 @@ func TestMissingConsortiumValue(t *testing.T) {
+	configTxDest := filepath.Join(tmpDir, "configtx")
+
+	config := genesisconfig.Load(genesisconfig.SampleSingleMSPChannelProfile, configtest.GetDevConfigDir())
+	ioutil.WriteFile(configTxDest, []byte{}, 0440)
```
Just to be safe, I think we should check the error here.

wlahti commented on cmd/configtxgen/main_test.go:
```diff
@@ -77,7 +78,20 @@ func TestMissingConsortiumValue(t *testing.T) {
+	assert.Error(t, doOutputChannelCreateTx(config, nil, "foo", configTxDest), "Not able to write to file")
+	assert.Regexp(t, regexp.MustCompile("Error writing channel create tx: open /tmp/configtxgen.+/configtx: permission denied"), err)
+
+	os.Remove(configTxDest)
```
I think we should use a defer for this remove call near where we create the file. That's the pattern we typically use in fabric.

wlahti reviewed (changes requested):
Thanks for taking the initiative to improve the tests here. Much appreciated! 

One important requested change (the regexp for the path doesn't work on all operating systems) and a few nits.

betasve commented:
Pushed some changes I think (as far as I understood your points) are addressing the concerns. Please, let me know if I missed something or something else needs improvement.

wlahti reviewed (changes requested):
Sorry, maybe should've been clearer in the OS-specific nature of the test that fails locally. The temp directory is different depending on the developer's OS, so we need to generalize the regular expression to not care about the beginning of the path. For reference, here's the output when running that test on my mac: 

```
=== RUN   TestUnsuccessfulChannelTxFileCreation
2020-04-14 16:30:43.549 EDT [common.tools.configtxgen.localconfig] Load -> INFO 01b Loaded configuration: /Users/wlahti/gopath/src/github.com/hyperledger/fabric/sampleconfig/configtx.yaml
2020-04-14 16:30:43.549 EDT [common.tools.configtxgen] doOutputChannelCreateTx -> INFO 01c Generating new channel configtx
2020-04-14 16:30:43.551 EDT [common.tools.configtxgen] doOutputChannelCreateTx -> INFO 01d Writing new channel tx
2020-04-14 16:30:43.551 EDT [common.tools.configtxgen] doOutputChannelCreateTx -> INFO 01e Generating new channel configtx
2020-04-14 16:30:43.553 EDT [common.tools.configtxgen] doOutputChannelCreateTx -> INFO 01f Writing new channel tx
    TestUnsuccessfulChannelTxFileCreation: main_test.go:92: 
        	Error Trace:	main_test.go:92
        	Error:      	Expect "Error writing channel create tx: open /var/folders/gy/jb454xb90513cqzgvv7111hr0000gn/T/configtxgen143477252/configtx: permission denied" to match "Error writing channel create tx: open /tmp/configtxgen.+/configtx: permission denied"
        	Test:       	TestUnsuccessfulChannelTxFileCreation
--- FAIL: TestUnsuccessfulChannelTxFileCreation (0.03s)
```

betasve commented:
Yes, sorry, I didn't really understand your point but it totally makes sense. I should have thought of that. Pushed an update.

wlahti commented on cmd/configtxgen/main_test.go:
```diff
@@ -77,7 +77,19 @@ func TestMissingConsortiumValue(t *testing.T) {
+
+	err := doOutputChannelCreateTx(config, nil, "foo", configTxDest).Error()
+	assert.Error(t, doOutputChannelCreateTx(config, nil, "foo", configTxDest), "Not able to write to file")
+	assert.Regexp(t, "Error writing channel create tx: open .+/configtxgen.+/configtx: permission denied", err)
```
```suggestion
	assert.EqualError(t, doOutputChannelCreateTx(config, nil, "foo", configTxDest), fmt.Sprintf("Error writing channel create tx: open %s: permission denied", configTxDest))
```
Not trying to stall on this one, but considering I'm not a maintainer and can't merge this in anyways, can't we simplify this down to just one line and avoid the regular expression by formatting in the path since we have the variable with the exact path. Tried this out locally and it works for me. Seems more flexible to me in case somebody down the line decided to change the name of the tmpDir.

betasve replied:
Sure, sure, I totally got your point. Pushing a change now. Let me know about the new version.

wlahti reviewed (dismissed):
Looks good to me now, thanks!

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -34,7 +35,7 @@ func TestMain(m *testing.M) {
 
 func TestInspectMissing(t *testing.T) {
-	assert.Error(t, doInspectBlock("NonSenseBlockFileThatDoesn'tActuallyExist"), "Missing block")
+	assert.EqualError(t, doInspectBlock("NonSenseBlockFileThatDoesn'tActuallyExist"), "Could not read block NonSenseBlockFileThatDoesn'tActuallyExist")
```
I know you didn't introduce these error messages, but per the general Go programming conventions, error strings should not start capitalized https://github.com/golang/go/wiki/CodeReviewComments#error-strings would you be able to go through the production code as well as the tests in this path and update the error strings accordingly? There's a lot of places in Fabric where we don't properly follow this convention but might as well start here since we're touching this path.

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -49,8 +50,8 @@ func TestInspectBlock(t *testing.T) {
 
-	assert.Error(t, doOutputBlock(config, "foo", ""), "Error writing genesis block:")
-	assert.Error(t, doInspectBlock(""), "Could not read block")
+	assert.EqualError(t, doOutputBlock(config, "foo", ""), "Error writing genesis block: open : no such file or directory")
```
Same here

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -59,7 +60,7 @@ func TestMissingOrdererSection(t *testing.T) {
 	config.Orderer = nil
 
-	assert.Error(t, doOutputBlock(config, "foo", blockDest), "Missing orderer section")
+	assert.EqualError(t, doOutputBlock(config, "foo", blockDest), "refusing to generate block which is missing orderer section")
```
Interestingly enough we follow the convention here, further demonstrating how inconsistent error handling is in Fabric

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -49,8 +50,8 @@ func TestInspectBlock(t *testing.T) {
-	assert.Error(t, doOutputBlock(config, "foo", ""), "Error writing genesis block:")
-	assert.Error(t, doInspectBlock(""), "Could not read block")
+	assert.EqualError(t, doOutputBlock(config, "foo", ""), "Error writing genesis block: open : no such file or directory")
+	assert.EqualError(t, doInspectBlock(""), "Could not read block ")
```
And here

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -77,7 +78,17 @@ func TestMissingConsortiumValue(t *testing.T) {
+	assert.NoError(t, ioutil.WriteFile(configTxDest, []byte{}, 0440))
+	defer os.Remove(configTxDest)
+
+	assert.EqualError(t, doOutputChannelCreateTx(config, nil, "foo", configTxDest), fmt.Sprintf("Error writing channel create tx: open %s: permission denied", configTxDest))
```
And here

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -115,15 +126,15 @@ func TestBadAnchorPeersUpdates(t *testing.T) {
 	config := genesisconfig.Load(genesisconfig.SampleSingleMSPChannelProfile, configtest.GetDevConfigDir())
 
-	assert.Error(t, doOutputAnchorPeersUpdate(config, "foo", configTxDest, ""), "Bad anchorPeerUpdate request - asOrg empty")
+	assert.EqualError(t, doOutputAnchorPeersUpdate(config, "foo", configTxDest, ""), "Must specify an organization to update the anchor peer for")
```
And here

caod123 commented on cmd/configtxgen/main_test.go:
```diff
@@ -115,15 +126,15 @@ func TestBadAnchorPeersUpdates(t *testing.T) {
 	backupApplication := config.Application
 	config.Application = nil
-	assert.Error(t, doOutputAnchorPeersUpdate(config, "foo", configTxDest, genesisconfig.SampleOrgName), "Bad anchorPeerUpdate request")
+	assert.EqualError(t, doOutputAnchorPeersUpdate(config, "foo", configTxDest, genesisconfig.SampleOrgName), "Cannot update anchor peers without an application section")
```
And here

caod123 reviewed (changes requested):
Some general comments on error string conventions

lindluni commented:
/azp run

wlahti commented:
Can you rebase this on master and we'll see if we can get a maintainer to merge as-is? I think the capitalization of the error messages can be done in a follow-up PR considering they've had that capitalization for years.

caod123 commented:
Looks like there aren't any conflicts with master, merging as is
